### PR TITLE
Add localState ngrxDirective option

### PR DESCRIPTION
**Pull request by mucaho**
@MrWolfZ 
As discussed in issue #165 , this PR adds:
* The `ngrxSuppressGlobalActionEmit` input boolean to ngrx's directives, which controls whether to dispatch form actions to the global `ActionsSubject` or not, `false` by default of course to not change current semantics. I thought this was better wording than `createLocalFormState`, but I can adapt this of course.
* The `ngrxFormAction` output event emitter that **always** emits any form action. It does not hurt if the action is always emitted, regardless of `ngrxSuppressGlobalActionEmit`, right?
* Unit and e2e tests
* A representative example showcasing the benefit of using local state. Showcases how to handle state update on inter-dependent form state and additional local state used in the component. Showcases how to properly detect changes in lieu of input properties, but still being as efficient as possible.
* Does **not** include any documentation. To be frank, I do not know where to put and what to write what's not already showcased in example. Can change this of course, depending on your input.

I hope everything checks out, otherwise let me know.

(e: Improved upon the example and some wording, pushed again)

### CODE REVIEW

**Comment by MrWolfZ** [comment]
Thanks a lot for this thorough PR. Overall it looks great. Here's a couple of comments:

> The `ngrxSuppressGlobalActionEmit` input boolean to ngrx's directives, which controls whether to dispatch form actions to the global `ActionsSubject` or not, false by default of course to not change current semantics. I thought this was better wording than `createLocalFormState`, but I can adapt this of course.

I like this design because it keeps the whole local form state concept in the directive without bleeding into the rest of **ngrx-forms**. At the same time I dislike its verbosity, because you need to pass the parameter to every single form control (which also makes it easy to accidentally forget adding it on a control). If there was an `isLocal` property on the form state then the directive could decide based on that, but this would require some changes to other areas. Thinking about it, there may also be a slightly different API design that at least takes care of the verbosity issue: instead of passing an additional flag there could be a new directive that uses `ngrxLocalFormControlState` (or `ngrxFormLocalControlState`) as a selector. This design still has the issue of it being possible to accidentally mix local and global form states though. What do you think?

> The `ngrxFormAction` output event emitter that always emits any form action. It does not hurt if the action is always emitted, regardless of `ngrxSuppressGlobalActionEmit`, right?

Yeah, its fine to always emit locally.

> A representative example showcasing the benefit of using local state. Showcases how to handle state update on inter-dependent form state and additional local state used in the component. Showcases how to properly detect changes in lieu of input properties, but still being as efficient as possible.

I feel that the example is maybe a bit too advanced to make the feature easy to understand. There should be at least one basic example that showcases purely how to handle only local form state without any interaction with effects etc, i.e. showing that this feature allows using **ngrx-forms** without ngrx. There could then be a second more advanced example for how to connect a local form with actions from the outside.

> Does not include any documentation. To be frank, I do not know where to put and what to write what's not already showcased in example. Can change this of course, depending on your input.

Yeah, a good example is always important. Even though the text documentation may not contain any extra details, it is still useful, since there are different people out there that learn through different things. Some may prefer looking at an example, others may prefer reading through documentation. For this feature we could probably just add a new markdown file `local-forms.md` to the `user-guide` section.

Overall, great work. Let's see if we can nail down the API design. While it's great that the PR is already so complete, in the future I would recommend to not start writing any documentation/examples before the design is finalized to save yourself some unnecessary refactoring/rewriting.

File: example-app/src/app/local-state/local-state.component.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      if (updated) {
+        // since OnPush is used, need to trigger detectChanges
+        // when action from outside updates localState
+        this.cd.detectChanges();
+      }
+    });
+  }
+
+  ngOnDestroy(): void {
+    this.subscription.unsubscribe();
```
Comment (MrWolfZ) [inline example-app/src/app/local-state/local-state.component.ts]: a tiny detail I like to keep in mind here is that `ngOnDestroy` can be called without `ngOnInit`, and therefore `this.subscription` can be undefined. You can either surround this with an `if` or initialize the variable upon declaration.

File: example-app/src/app/local-state/local-state.reducer.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+export const INITIAL_LOCAL_STATE: LocalState = {
+  manufacturers: [],
+  formState: INITIAL_FORM_STATE
+};
+
+const reducers = combineReducers<LocalState>({
+  manufacturers(manufacturers = [], a: Action) {
+    // update from loaded data
+    if (a instanceof SetManufacturersAction) {
```
Comment (MrWolfZ) [inline example-app/src/app/local-state/local-state.reducer.ts]: personally I never use `instanceof` with actions, but instead compare the `type`. This is because it is not guaranteed the action will always be a class instance (e.g. if dispatched via the redux devtools).

File: src/control/directive.spec.ts [hunk 3]
```diff
@@ -560,6 +560,113 @@ describe(NgrxFormControlDirective.name, () => {
...
   });
 
+  describe('suppression of global action emit', () => {
+    it('be disabled by default', () => {
+      expect(directive.ngrxSuppressGlobalActionEmit).toBe(false);
+      directive.ngOnInit();
+      expect(directive.ngrxSuppressGlobalActionEmit).toBe(false);
+    });
+
+    describe('is disabled', () => {
```
Comment (MrWolfZ) [inline src/control/directive.spec.ts]: aren't these tests redundant since they represent the existing behavior that is already tested?

File: src/control/directive.ts [hunk 4]
```diff
@@ -86,6 +88,10 @@ export class NgrxFormControlDirective<TStateValue extends FormControlValueTypes,
 
   @Input() ngrxValueConverter: NgrxValueConverter<TViewValue, TStateValue> = NgrxValueConverters.default<any>();
 
+  @Input() ngrxSuppressGlobalActionEmit: boolean = false;
```
Comment (MrWolfZ) [inline src/control/directive.ts]: minor thing but the type annotation is redundant. i'd remove it for consistency with the rest of the code

File: src/control/e2e.spec/select.spec.ts [hunk 5]
```diff
@@ -172,3 +172,70 @@ describe(NumberSelectComponent.name, () => {
     element.dispatchEvent(new Event('change'));
   });
 });
+
+@Component({
```
Comment (MrWolfZ) [inline src/control/e2e.spec/select.spec.ts]: the reason for `e2e.spec/select.spec.ts` is that there were some tricky interaction issues with angular itself. This is unrelated to the local form state. It's still a good idea to have an end to end test though, but I'd put it in a separate `local-state.spec.ts` file and maybe use just an `input` instead of the more complex `select`

File: src/group/directive.spec.ts [hunk 6]
```diff
@@ -46,4 +46,107 @@ describe(NgrxFormDirective.name, () => {
     directive.onSubmit({ preventDefault: () => void 0 } as any);
     actionsSubject.complete();
   });
+
+  describe('suppression of global action emit', () => {
+    it('be disabled by default', () => {
```
Comment (MrWolfZ) [inline src/group/directive.spec.ts]: missing `should`?

File: src/group/directive.spec.ts [hunk 7]
```diff
@@ -46,4 +46,107 @@ describe(NgrxFormDirective.name, () => {
...
+      expect(directive.ngrxSuppressGlobalActionEmit).toBe(false);
+      directive.ngOnInit();
+      expect(directive.ngrxSuppressGlobalActionEmit).toBe(false);
+    });
+
+    describe('is disabled', () => {
```
Comment (MrWolfZ) [inline src/group/directive.spec.ts]: same as for control directive, I think these tests are redundant

**Comment by mucaho** [comment]
Thanks for the review, i will incorporate all the remarks once we settle on the API, as you said.

One thing I am not certain about: For now we have the `ngrxFormAction` output on both `ngrxFormState` and `ngrxFormControlState`. Should the output be called `ngrxFormControlAction` in the latter?

 > At the same time I dislike _[the `ngrxSuppressGlobalActionEmit` input boolean's]_  verbosity, because you need to pass the parameter to every single form control (which also makes it easy to accidentally forget adding it on a control). If there was an `isLocal` property on the form state then the directive could decide based on that, but this would require some changes to other areas. Thinking about it, there may also be a slightly different API design that at least takes care of the verbosity issue: instead of passing an additional flag there could be a new directive that uses `ngrxLocalFormControlState` (or `ngrxFormLocalControlState`) as a selector. This design still has the issue of it being possible to accidentally mix local and global form states though. What do you think?

* The `isLocal` form state property would be the least verbose one, and once you have set it, you would not have to think about using the proper input / directive for all your form controls. While this sounds nice I'm not a big fan of mixing form state with configuration options.
*  `ngrxLocalFormControlState` has the same issue of being accidentally mixed up or forgotten about like `ngrxSuppressGlobalActionEmit`. However, it's less verbose as you eliminate the need for this additional input.

In my opinion another set of directives (`ngrxLocalFormControlState` & ``ngrxLocalFormState`) would be the best option.

If you agree, I would proceed with this approach. However, in this case the question remains how to go about the implementation. As far as I can tell, there are two options:
* make `ngrxLocalFormControlState` directive extend `ngrxFormControlState` directive, passing the former's output `EventEmitter` in the super constructor as the `ActionsSubject`. A bit hacky since `EventEmitter` being a subject is an implementation detail, which theoretically could change in future releases.
* rename current `ngrxFormControlState` directive to `ngrxLocalFormControlState`. Remove `ActionsSubject` from constructor and only emit to the local `ngrxFormAction` output.
Make a new directive called `ngrxFormControlState` which has the same constructor parameters plus the `ActionsSubject` and overrides the `dispatchAction` method by also emitting to this `ActionsSubject`. (And hope for the best that angular inherits all decorators properly).

**Comment by MrWolfZ** [comment]
> One thing I am not certain about: For now we have the ngrxFormAction output on both ngrxFormState and ngrxFormControlState. Should the output be called ngrxFormControlAction in the latter?

Hmm, I am honestly divided. In the grand scheme of things this is probably just a minor detail though, so any option is fine. We could also just call it `ngrxFormsAction` to indicate these are actions from this library. Choose whatever feels right to you.

> In my opinion another set of directives (`ngrxLocalFormControlState` & `ngrxLocalFormState`) would be the best option.

Yeah, let's go with this design.

> If you agree, I would proceed with this approach. However, in this case the question remains how to go about the implementation. As far as I can tell, there are two options:

I was thinking about this as well. I would go for the inheritance approach. You can simply make the `ActionsSubject` constructor parameter `@Optional()` and then provide `null` as a value to the super class constructor in the new directive. In addition you would keep the `dispatchAction` method and just override it in the subclass to emit on the local `EventEmitter` instead of the `ActionsSubject`. This should work for both directives.

**Comment by mucaho** [comment]
I encountered some unexpected roadblocks while trying to implement the directive `ngrxLocalFormControlState` as a subclass of `ngrxFormControlState`:
* My e2e test kept failing, much to my surprise. Reason is that the new directive `ngrxLocalFormControlState` can not find any `@Self`-scoped `FormViewAdapter`s. After some investigation I found out that these are added by the same setter property `ngrxFormControlState` in the various adapters. While adding another selector and input setter would be possible for in-house viewAdapters that same can not be said for external adapters that exist already and rely on this naming convention. e.g. for `NgrxRadioViewAdapter`:
```ts
@Directive({
  selector: 'input[type=radio][ngrxFormControlState]',
  // ...
})
export class NgrxRadioViewAdapter implements FormViewAdapter, OnInit, AfterViewInit {
  @Input() set ngrxFormControlState(value: FormControlState<any>) { /*...*/ }
}
```
* The documentation section on [user-defined properties](https://ngrx-forms.readthedocs.io/en/master/user-guide/form-controls/#user-defined-properties) uses the same, aforementioned naming convention. Same problem again.
* Same problem for `CustomErrorStateMatcherDirective`, although I don't know for what that's used.

Here the implementation for reference:
```ts
@Directive({
  // tslint:disable-next-line:directive-selector
  selector: '[ngrxLocalFormControlState]',
})
export class NgrxLocalFormControlDirective<TStateValue extends FormControlValueTypes, TViewValue = TStateValue>
  extends NgrxFormControlDirective<TStateValue, TViewValue> {

  @Input() set ngrxLocalFormControlState(state: FormControlState<TStateValue>) {
    this.ngrxFormControlState = state;
  }

  @Output() ngrxFormsAction = new EventEmitter<Actions<TStateValue>>();

  constructor(
    el: ElementRef,
    @Optional() @Inject(DOCUMENT) dom: Document | null,
    @Self() @Optional() @Inject(NGRX_FORM_VIEW_ADAPTER) viewAdapters: FormViewAdapter[],
    @Self() @Optional() @Inject(NG_VALUE_ACCESSOR) valueAccessors: ControlValueAccessor[],
  ) {
    super(el, dom, null as any as ActionsSubject, viewAdapters, valueAccessors);
  }

  protected dispatchAction(action: Actions<TStateValue>) {
    this.ngrxFormsAction.emit(action);
  };

}
```

-----

If you know of a good way of introducing another directive, while still being backward-compatible let me know. A cool solution would be to apply the new directive on all elements that have the output binding too, something like `selector: [ngrxFormControlState] (ngrxFormsAction)`, but that does not work of course.

Otherwise, maybe we should go ahead with the original approach. Looking forward to hear your thoughts.

File: src/control/e2e.spec/select.spec.ts (hunk 5 above)
Comment (mucaho) [reply src/control/e2e.spec/select.spec.ts]: Was not able to trigger a simple change event manually on an input element and have ngrx forms register it properly. Extracted the example into a separate spec file though.

**Comment by mucaho** [comment]
Changes from previous version:
* Added the local directives. as mentioned in previous comment they are added if you bind an event listener to `(ngrxFormsAction)`, otherwise the default directives are used.
* Add an introductory local state example.
* Added documentation.

Codecov is failing because I did not test the branches where the `@Optional`ly injected `ActionsSubject` is `null`. Should I throw an error in this case? It should never happen, as I override the `dispatchMethod` in that case when it's null.

**Comment by MrWolfZ** [comment]
> Scratch everything I wrote before, I can make the new directive's selector be `[ngrxFormControlState][ngrxFormsAction]` and change the old one's to `[ngrxFormControlState]:not([ngrxFormsAction])` and everything looks to be working properly.

Yeah, that's a pretty cool approach. For reference in the future, I believe you can also have multiple selectors for a component, e.g `[ngrxFormControlState],[ngrxLocalFormControlState]`. However, I believe your solution is much more elegant since it is much less invasive than using a different attribute as a selector. I'll take a look at the code.

File: docs/user-guide/local-form-state.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+@Component({
+  selector: 'my-local-form-component',
+  template: `
+    <form>
+      <input type="text"
+        [ngrxFormControlState]="formState.controls.myLocalFormField"
+        (ngrxFormsAction)="handleFormAction($event)">
+    </form>
+  `,
+  changeDetection: ChangeDetectionStrategy.OnPush,
```
Comment (MrWolfZ) [inline docs/user-guide/local-form-state.md]: I'd remove this from the example.

File: docs/user-guide/local-form-state.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  providers: [],
+  bootstrap: [AppComponent]
+})
+export class AppModule { }
+```
+
+Maintaining a local form state has its merits:
+
+* The root or feature store is not populated with **temporary form data** which perishes after the component is destroyed.
+Form data is often used for creating and or updating an entity and becomes obsolete after an update request is sent to the server. After the server responds, you can update your list or table of entities with the new or updated, proper entity data.
```
Comment (MrWolfZ) [inline docs/user-guide/local-form-state.md]: typo in `creating and or updating`? If you want both, separate them with a slash, i.e. `creating and/or updating`.

File: example-app/src/app/local-state-advanced/local-state-advanced.component.html [hunk 10]
```diff
@@ -0,0 +1,52 @@
+<ngf-form-example exampleName="Local state: Advanced"
+                  [formState]="localState"
+                  >
+  Advanced example that shows how you can disable emitting any form actions to the global store via the builtin ActionSubject.
```
Comment (MrWolfZ) [inline example-app/src/app/local-state-advanced/local-state-advanced.component.html]: typo `ActionSubject` -> `ActionsSubject`

File: example-app/src/app/local-state-advanced/local-state-advanced.component.html [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  Instead, you can listen for actions using the directive's output EventEmitter and route them to the component.
+  <br />
+  The component holds, updates and maintains a local state, which does not taint the global store and
+  is automatically lost once the component is destroyed.
```
Comment (MrWolfZ) [inline example-app/src/app/local-state-advanced/local-state-advanced.component.html]: `lost` sounds quite negative, I would use `cleaned up` or `removed`

File: example-app/src/app/local-state-advanced/local-state-advanced.component.ts [hunk 12]
```diff
@@ -0,0 +1,55 @@
+import { ChangeDetectionStrategy, Component, ChangeDetectorRef, OnDestroy, OnInit } from '@angular/core';
+import { ActionsSubject, Action } from '@ngrx/store';
+import { SetValueAction, Actions } from 'ngrx-forms';
+import { Subscription } from 'rxjs';
+
+import { reducer, LocalState, GetManufacturersAction, INITIAL_LOCAL_STATE, FORM_ID } from './local-state-advanced.reducer';
+
+
```
Comment (MrWolfZ) [inline example-app/src/app/local-state-advanced/local-state-advanced.component.ts]: remove this empty line

File: example-app/src/app/local-state-advanced/local-state-advanced.component.ts [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+@Component({
+  selector: 'ngf-local-state-advanced',
+  templateUrl: './local-state-advanced.component.html',
+  styleUrls: ['./local-state-advanced.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class LocalStateAdvancedComponent implements OnInit, OnDestroy {
+
+  public localState: LocalState = INITIAL_LOCAL_STATE;
```
Comment (MrWolfZ) [inline example-app/src/app/local-state-advanced/local-state-advanced.component.ts]: `public` and type declaration are redundant

File: example-app/src/app/local-state-advanced/local-state-advanced.component.ts [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+  ngOnDestroy(): void {
+    this.subscription.unsubscribe();
+  }
+
+  handleFormAction(action: Actions<any>) {
+    this.updateState(action);
+
+    // trigger loading of new manufacturers list in effect
+    if (action.type === SetValueAction.TYPE && action.controlId === `${FORM_ID}.countryCode`) {
```
Comment (MrWolfZ) [inline example-app/src/app/local-state-advanced/local-state-advanced.component.ts]: Instead of `${FORM_ID}.countryCode` which is not refactoring safe, I would just use `this.localState.formState.controls.CountryCode.id`, even if it is longer

File: example-app/src/app/local-state-advanced/local-state-advanced.reducer.ts [hunk 15]
```diff
@@ -0,0 +1,74 @@
+import { Action, combineReducers } from '@ngrx/store';
+import { createFormGroupState, formGroupReducer, FormGroupState, updateGroup, setValue } from 'ngrx-forms';
+
+export class GetManufacturersAction implements Action {
+  static readonly TYPE = 'localState/GET_MANUFACTURERS';
```
Comment (MrWolfZ) [inline example-app/src/app/local-state-advanced/local-state-advanced.reducer.ts]: change type prefix to `localStateAdvanced`

File: example-app/src/app/local-state-introduction/local-state-introduction.component.html [hunk 16]
```diff
@@ -0,0 +1,26 @@
+<ngf-form-example exampleName="Local state: Introduction"
+                  [formState]="formState"
+                  >
+  Introductory example that shows how you can disable emitting any form actions to the global store via the builtin ActionSubject.
```
Comment (MrWolfZ) [inline example-app/src/app/local-state-introduction/local-state-introduction.component.html]: typo `ActionSubject` -> `ActionsSubject`

File: example-app/src/app/local-state-introduction/local-state-introduction.component.html [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+  Instead, you can listen for actions using the directive's output EventEmitter and route them to the component.
+  <br />
+  The component holds, updates and maintains a local form state, which does not taint the global store and
+  is automatically lost once the component is destroyed.
```
Comment (MrWolfZ) [inline example-app/src/app/local-state-introduction/local-state-introduction.component.html]: `cleaned up` or `removed` instead of `lost`

File: example-app/src/app/local-state-introduction/local-state-introduction.component.ts [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+import { Actions, FormGroupState } from 'ngrx-forms';
+
+import { reducer, INITIAL_FORM_STATE, FormValue } from './local-state-introduction.reducer';
+
+
+@Component({
+  selector: 'ngf-local-state-introduction',
+  templateUrl: './local-state-introduction.component.html',
+  styleUrls: ['./local-state-introduction.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
```
Comment (MrWolfZ) [inline example-app/src/app/local-state-introduction/local-state-introduction.component.ts]: remove `changeDetection`

File: example-app/src/app/local-state-introduction/local-state-introduction.component.ts [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+})
+export class LocalStateIntroductionComponent {
+
+  public formState: FormGroupState<FormValue> = INITIAL_FORM_STATE;
```
Comment (MrWolfZ) [inline example-app/src/app/local-state-introduction/local-state-introduction.component.ts]: `public` and type declaration are redundant

File: example-app/src/app/local-state-introduction/local-state-introduction.reducer.ts [hunk 20]
```diff
@@ -0,0 +1,17 @@
+import { Action } from '@ngrx/store';
+import { createFormGroupState, formGroupReducer, FormGroupState } from 'ngrx-forms';
+
+
```
Comment (MrWolfZ) [inline example-app/src/app/local-state-introduction/local-state-introduction.reducer.ts]: remove extra blank line

File: src/control/directive.ts [hunk 21]
```diff
@@ -53,7 +53,7 @@ class ControlValueAccessorAdapter implements FormViewAdapter {
 
 @Directive({
   // tslint:disable-next-line:directive-selector
-  selector: '[ngrxFormControlState]',
+  selector: ':not([ngrxFormsAction])[ngrxFormControlState]',
```
Comment (MrWolfZ) [inline src/control/directive.ts]: i'd prefer `[ngrxFormControlState]:not([ngrxFormsAction])`

File: src/control/directive.ts [hunk 22]
```diff
@@ -112,7 +112,7 @@ export class NgrxFormControlDirective<TStateValue extends FormControlValueTypes,
     // for the dom parameter the `null` type must be last to ensure that in the compiled output
     // there is no reference to the Document type to support non-browser platforms
     @Optional() @Inject(DOCUMENT) private dom: Document | null,
-    private actionsSubject: ActionsSubject,
+    @Optional() private actionsSubject: ActionsSubject,
```
Comment (MrWolfZ) [inline src/control/directive.ts]: if it is optional, then you should type it as `ActionsSubject | null` since Angular provides `null` for non-existing optional dependencies. This will ensure you are not accidentally accessing the subject anywhere unsafely

File: src/control/e2e.spec/local-state.spec.ts [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+import { Action, ActionsSubject } from '@ngrx/store';
+import { Observable, Subject } from 'rxjs';
+import { count } from 'rxjs/operators';
+
+import { MarkAsDirtyAction } from '../../actions';
+import { NgrxFormsModule } from '../../module';
+import { createFormControlState, FormControlState } from '../../state';
+
+
```
Comment (MrWolfZ) [inline src/control/e2e.spec/local-state.spec.ts]: remove extra blank line

File: src/control/e2e.spec/local-state.spec.ts [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+  beforeEach(() => {
+    fixture = TestBed.createComponent(NumberSelectComponentLocalState);
+    component = fixture.componentInstance;
+    component.state = INITIAL_STATE;
+    fixture.detectChanges();
+    const nativeElement = fixture.nativeElement as HTMLElement;
+    element = nativeElement.querySelector('select')!;
+  });
+
+  it(`should not trigger a ${MarkAsDirtyAction.name}to the global store when an option is selected`, done => {
```
Comment (MrWolfZ) [inline src/control/e2e.spec/local-state.spec.ts]: missing space before `to`

File: src/control/local-state-directive.ts [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+} from '@angular/core';
+import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
+import { ActionsSubject } from '@ngrx/store';
+
+import { FormControlValueTypes } from '../state';
+import { FormViewAdapter, NGRX_FORM_VIEW_ADAPTER } from '../view-adapter/view-adapter';
+import { NgrxFormControlDirective, Document } from './directive';
+import { Actions } from '../actions';
+
+
```
Comment (MrWolfZ) [inline src/control/local-state-directive.ts]: remove extra blank line

File: src/control/local-state-directive.ts [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+  @Output() ngrxFormsAction = new EventEmitter<Actions<TStateValue>>();
+
+  constructor(
+    el: ElementRef,
+    @Optional() @Inject(DOCUMENT) dom: Document | null,
+    @Self() @Optional() @Inject(NGRX_FORM_VIEW_ADAPTER) viewAdapters: FormViewAdapter[],
+    @Self() @Optional() @Inject(NG_VALUE_ACCESSOR) valueAccessors: ControlValueAccessor[],
+  ) {
+    super(el, dom, null as any as ActionsSubject, viewAdapters, valueAccessors);
```
Comment (MrWolfZ) [inline src/control/local-state-directive.ts]: you can remove the casts when the base class constructor accepts `null`

File: src/control/local-state-directive.ts [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  protected dispatchAction(action: Actions<TStateValue>) {
+    this.ngrxFormsAction.emit(action);
+  };
+
```
Comment (MrWolfZ) [inline src/control/local-state-directive.ts]: remove extra blank line

File: src/group/directive.ts [hunk 28]
```diff
@@ -11,13 +11,19 @@ interface CustomEvent extends Event { }
 
 @Directive({
   // tslint:disable-next-line:directive-selector
-  selector: 'form[ngrxFormState]',
+  selector: 'form:not([ngrxFormsAction])[ngrxFormState]',
```
Comment (MrWolfZ) [inline src/group/directive.ts]: i'd prefer `form[ngrxFormState]:not([ngrxFormsAction])`

File: src/group/directive.ts [hunk 29]
```diff
@@ -11,13 +11,19 @@ interface CustomEvent extends Event { }
...
 })
 export class NgrxFormDirective<TValue extends { [key: string]: any }> implements OnInit {
   // tslint:disable-next-line:no-input-rename
   @Input('ngrxFormState') state: FormGroupState<TValue>;
 
-  constructor(private actionsSubject: ActionsSubject) { }
+  constructor(@Optional() private actionsSubject: ActionsSubject) { }
```
Comment (MrWolfZ) [inline src/group/directive.ts]: type as `ActionsSubject | null`

File: src/group/local-state-directive.ts [hunk 30]
```diff
@@ -0,0 +1,24 @@
...
+  // tslint:disable-next-line:directive-selector
+  selector: 'form[ngrxFormState][ngrxFormsAction]',
+})
+export class NgrxLocalFormDirective<TValue extends { [key: string]: any }>
+  extends NgrxFormDirective<TValue> {
+
+  @Output() ngrxFormsAction = new EventEmitter<Actions<TValue>>();
+
+  constructor() {
+    super(null as any as ActionsSubject);
```
Comment (MrWolfZ) [inline src/group/local-state-directive.ts]: can remove cast when base constructor accepts `null`

File: src/control/directive.ts (hunk 22 above)
Comment (mucaho) [reply src/control/directive.ts]: That is not possible due to how the Angular DI mechanism works. If you extend it to union type `ActionsSubject | null` it will complain that it can not find the required `InjectionToken`. I have introduced a private field with the proper type declaration and just set it the parameter to the field.

File: src/group/directive.ts (hunk 28 above)
Comment (mucaho) [reply src/group/directive.ts]: Tried that initially, but the selector won't work this way.

File: src/group/directive.ts (hunk 29 above)
Comment (mucaho) [reply src/group/directive.ts]: See https://github.com/MrWolfZ/ngrx-forms/pull/166#discussion_r349931342

File: src/control/directive.ts (hunk 21 above)
Comment (mucaho) [reply src/control/directive.ts]: See https://github.com/MrWolfZ/ngrx-forms/pull/166#discussion_r349931342

File: src/control/local-state-directive.ts (hunk 26 above)
Comment (mucaho) [reply src/control/local-state-directive.ts]: See https://github.com/MrWolfZ/ngrx-forms/pull/166#discussion_r349931342

File: src/group/local-state-directive.ts (hunk 30 above)
Comment (mucaho) [reply src/group/local-state-directive.ts]: See https://github.com/MrWolfZ/ngrx-forms/pull/166#discussion_r349931342

**Comment by mucaho** [comment]
Alright, applied all proposed changes, or left responses where it's not possible to do so.
I hope we are good to go now.

File: src/control/directive.ts (hunk 22 above)
Comment (MrWolfZ) [reply src/control/directive.ts]: Hmm, I believe it would work if you specify `@Inject(ActionsSubject)` but I am honestly not 100% sure. I'll look into this myself after merging.

**Comment by MrWolfZ** [comment]
Yup, we are good.

**Comment by MrWolfZ** [comment]
And it's published as version [6.1.0](https://github.com/MrWolfZ/ngrx-forms/releases/tag/v6.1.0).

File: src/control/directive.ts (hunk 22 above)
Comment (mucaho) [reply src/control/directive.ts]: Ah, I see, good to know that it can be done more elegantly with this.

**Comment by mucaho** [comment]
Thanks for all your input, I am glad we managed to find a nice API and provide it with examples & documentation!

I see you have done some linting afterwards, thanks!
As you have mentioned already before, I think you can add `tslint` to the `npm test` script, so the CI check fails if it's not properly formatted. This would remove the need for the manual chore work of fixing formatting, as you can run `tslint --fix` to resolve most of the problems.
